docs/derivation-rules.md

```diff
@@ -0,0 +1,71 @@
+This file shows derivation rules used in fialyzer.
+
+Our derivation rules are almost same as [the original paper](https://it.uu.se/research/group/hipe/papers/succ_types.pdf)'s one, but extended by remote call, local call, list, etc.
+
+```
+A : context (mapping of variable to type)
+τ : type
+α, β : type variable
+x : variable
+e : term
+fun f/a : local function (f: function name, a: arity)
+fun m:f/a : remote function (m: module name, f: function name, a: arity)
+```
+
+```
+-------------------------- [VAR]
+A ∪ {x → τ} ⊢ x : τ, ∅
+
+A ⊢ e1 : τ1, C1 ... en : τn, Cn
+------------------------------------------------------ [STRUCT]
+A ⊢ {e1, ... , en} : {τ1, ... , τn}, C1 ∧ ... ∧ Cn
+
+A ⊢ e1 : τ1, C1         A ∪ {x → τ1} ⊢ e2 : τ2, C2
+---------------------------------------------------- [LET]
+A ⊢ let x = e1 in e2 : τ2, C1 ∧ C2
+
+A ∪ {fun xi/ai → αi} ⊢ f1 : τ1, C1 ... fn : τn, Cn      e : τ, C   where ai = length (args fi)
+------------------------------------------------------------------------------------------------- [LETREC]
+A ⊢ letrec x1 = f1, ... , xn = fn in e : τ, C1 ∧ ... ∧ Cn ∧ C ∧ (τ1 = α1) ∧ ... ∧ (τn = αn)
+
+A ∪ {x1 → α1, ... , xn → αn} ⊢ e : τ, C
+---------------------------------------------------------------------- [ABS]
+A ⊢ fun(x1, ... , xn) → e : (α1, ... , αn) → τ, C
+
+A ⊢ e : τ, C  e1 : τ1, C1 ... en : τn, Cn
+-------------------------------------------------------------------------------------------------------------- [APP]
+A ⊢ e(e1, ... , en) : β, (τ = (α1, ... , αn) → α) ∧ (β ⊆ α) ∧ (τ1 ⊆ α1) ∧ ... ∧ (τn ⊆ αn) ∧ C ∧ C1 ∧ ... ∧ Cn
+
+A ⊢ p : τ, Cp     A ⊢ g : τg, Cg
+------------------------------------------ [PAT]
+A ⊢ p when g : τ, (τg ⊆ boolean()) ∧ Cp ∧ Cg
+
+                 A ∪ {v → αv | v ∈ Var(p1)} ⊢ p1 : τp1, Cp1,  b1 : τb1, Cb1
+                                           ...
+A ⊢ e : τ, Ce    A ∪ {v → αv | v ∈ Var(pn)} ⊢ pn : τpn, Cpn,  bn : τbn, Cbn
+-------------------------------------------------------------------------------------------------------------------- [CASE]
+A ⊢ case e of p1 → b1; ... pn → bn end : β, Ce ∧ (C1 ∨ ... ∨ Cn) where Ci = ((β = τbi) ∧ (τ = τpi) ∧ Cpi ∧ Cbi)
+
+------------------------------------- [LOCALFUN]
+A ∪ {fun f/a → τ} ⊢ fun f/a : τ, ∅
+
+----------------------------------------- if m and f is atom literal, a is non_neg_integer literal [MFA]
+A ∪ {fun m:f/a → τ} ⊢ fun m:f/a : τ, ∅
+
+A ⊢ m : τm, Cm   A ⊢ f : τf, Cf   A ⊢ a : τa, Ca
+-------------------------------------------------------------------------------------------------- if either m, f, a is not atom literal or non_neg_integer literal [MFAEXPR]
+A ⊢ fun m:f/a : τ, (τ ⊆ any()) ∧ (τm ⊆ atom()) ∧ (τf ⊆ atom()) ∧ (τa ⊆ number()) ∧ Cm ∧ Cf ∧ Ca
+```
+
+The following derivation rules are the differences from the original paper.
```

Comment:
How about it like the following?
```suggestion
The differences from the derivation rules on the original paper are as follows.
```

Reply:
I'll replace with your suggestion!

Reply:
fixed https://github.com/dwango/fialyzer/pull/164/commits/c211189783aab91a8fc6d4a4fcb8684837cdad39